doc/arch/020-questionnaires-aka-dynamic-forms.md

```diff
@@ -0,0 +1,127 @@
+# Questionnaires (AKA Dynamic Forms)
+
+## Status
+Accepted
+
+## Context
+
+Historically, a "form" at Simple has been static and designed as a first-class entity. For example, Blood Pressure has it's own API, database and a static UI on mobile. The challenge with static forms has been high time to production and inflexible UI. A new form demands a new API, table and UI development on mobile. Modifications to an existing form demands a version upgrade of the API and mobile app.
+
+Static forms have served us well for infrequently changing forms like Blood Pressure and Blood Sugar. For forms that require frequent changes (like Drug-stock), we used WebView to accommodate the dynamic nature of this form. WebView isn't offline-first, like the rest of Simple app.
+
+Questionnaires (aka Dynamic forms) is an effort at solving 2 challenges in an offline-first manner: reducing time to production & a flexibility to add/remove fields without need to update the Simple app. We also refer to them as Server-driven UI.
+
+## Decision
+
+We have designed Questionnaires by reusing underlying offline-first sync architecture of Simple. Definitions of new terminology introduced:
+- **Form:** A document to gather information in a structured manner.
+- **Questionnaire:** A sync resource containing form's layout and input fields, version and form-type.
+- **Questionnaire Response:** A sync resource containing user's inputs for a form/questionnaire, and user/facility the input was recorded at.
+- **Layout:** A custom DSL that helps Mobile rendering a form defined by server.
+
+## Design (AKA Implementation Details)
+
+Questionnaires are implemented by following below mentioned steps:
+
+### 1. Mobile _syncs_ questionnaires from server to device
+Mobile requests questionnaires from server using [the questionnaire sync API](https://api.simple.org/api-docs/index.html#tag/Questionnaires). Server responds with all layouts compatible with the DSL version supported by the mobile app.
+
+On Mobile database, the primary key for Questionnaires sync resource is questionnaire_type, and not id. _That is a key difference between Questionnaire and other sync resources._ Reason being, mobile needs only 1 layout per type of a questionnaire. Whenever a new questionnaire of an existing type is created, server sends the updated ID & layout in the sync response, and mobile updates the row for corresponding type.
+
+| questionnaire_type <PK> | id     | layout   |
+|-------------------------|--------|----------|
+| screening_reports       | uuid-1 | {form-1} |
+| supplies_reports        | uuid-2 | {form-2} |
+
+### 2. Mobile _syncs_ questionnaire_responses from server to device
+User inputs for a dynamic form are stored in a questionnaire response. Mobile requests them from server using [this sync API](https://api.simple.org/api-docs/index.html#tag/Questionnaire-Responses).
```

Comment:
This is for mobile syncing from server. Section # 6 & 7 talks about mobile sending questionnaires to server.

Reply:
Yeah. It's unclear because when you read that user inputs are stores in a response, you expect the server to fetch the response.

```suggestion
User inputs for a dynamic form are stored in a questionnaire response. Mobile requests response objects from the server using [this sync API](https://api.simple.org/api-docs/index.html#tag/Questionnaire-Responses) so they are available to the mobile user to store inputs.
```

Reply:
Understood. Reworded the sentence to use questionnaire response sync resource (instead of response objects).